src/flow_graph.py

```diff
@@ -656,10 +712,13 @@ def find_block_by_label(label: str) -> Optional[Block]:
         nodes.append(new_node)
 
         # Recursively analyze.
-        next_block = blocks[block.index + 1]
-        new_node.successor = build_graph_from_block(
-            next_block, blocks, nodes, asm_data, arch
-        )
+        if block.index + 1 < len(blocks):
+            next_block = blocks[block.index + 1]
+            new_node.successor = build_graph_from_block(
+                next_block, blocks, nodes, asm_data, arch
+            )
+        else:
+            new_node.successor = terminal_node
```

Comment:
> (I don't know any cases where that safeguard doesn't work)

```
glabel test
nop
.foo:
nop
b .foo
nop
jal bar
nop
```
should fail to trigger it, same with `beqz $a0, .foo` instead of `b .foo`

Reply:
I added a check to produce this with `beqz $a0, .foo` (the `b .foo` version produces the same C as master)


```
$ cat bad.s 
glabel test
nop
.foo:
nop
beqz $a0, .foo
nop
jal bar
nop
$ ./mips_to_c.py bad.s
/*
Decompilation failure in function test:

Missing return instruction in block: # 1 (foo)
    nop
    beqz $a0, .foo
    nop

# 1 -> cond: 2, def: 2
*/
```

Reply:
Nice. What's unfortunate though is that this is a hard error while the safeguard repairs the function and tries anyway, and also that we now have two separate ways of handling this kind of error... Here's a suggestion:
- add the safeguard `jr $ra` unconditionally
- check if it has in edges; if so, emit a warning
- now that the last block is guaranteed to always be a return node, remove the block.index bounds checking again, as well as the DecompFailure you added

This doesn't work for fragments, but those can instead add a safeguard empty block and when encountering that in `build_graph_from_block` return TerminalNode.

(this is getting complex for such a minor point, sorry...)

Reply:
> This doesn't work for fragments, but those can instead add a safeguard empty block and when encountering that in `build_graph_from_block` return TerminalNode.

I riffed on this and reverted most of the changes, but introduced this smaller change to `build_nodes` that fixes all the logic.
```python
    if fragment:
        blocks.append(terminal_node.block)
```

Reply:
That's a fair solution. You can back out the added DecompFailure too since it's now impossible to trigger. (I might experiment with my suggested improved safeguard later, would be nice to finally do that properly.)